src/librustdoc/docfs.rs

```diff
@@ -0,0 +1,78 @@
+//! Rustdoc's FileSystem abstraction module.
+//!
+//! On Windows this indirects IO into threads to work around performance issues
+//! with Defender (and other similar virus scanners that do blocking operations).
+//! On other platforms this is a thin shim to fs.
+//!
+//! Only calls needed to permit this workaround have been abstracted: thus
+//! fs::read is still done directly via the fs module; if in future rustdoc
+//! needs to read-after-write from a file, then it would be added to this
+//! abstraction.
+
+use std::fs;
+use std::io;
+use std::path::Path;
+
+macro_rules! try_err {
+    ($e:expr, $file:expr) => {{
+        match $e {
+            Ok(e) => e,
+            Err(e) => return Err(E::new(e, $file)),
+        }
+    }};
+}
+
+pub trait PathError {
+    fn new<P: AsRef<Path>>(e: io::Error, path: P) -> Self;
+}
+
+pub struct DocFS {
+    sync_only: bool,
+}
+
+impl DocFS {
+    pub fn new() -> DocFS {
+        DocFS {
+            sync_only: false,
+        }
+    }
+
+    pub fn set_sync_only(&mut self, sync_only: bool) {
+        self.sync_only = sync_only;
+    }
+
+    pub fn create_dir_all<P: AsRef<Path>>(&self, path: P) -> io::Result<()> {
+        // For now, dir creation isn't a huge time consideration, do it
+        // synchronously, which avoids needing ordering between write() actions
+        // and directory creation.
+        fs::create_dir_all(path)
+    }
+
+    pub fn write<P, C, E>(&self, path: P, contents: C) -> Result<(), E>
+    where
+        P: AsRef<Path>,
+        C: AsRef<[u8]>,
+        E: PathError,
+    {
+        if !self.sync_only && cfg!(windows) {
+            // A possible future enhancement after more detailed profiling would
+            // be to create the file sync so errors are reported eagerly.
+            let contents = contents.as_ref().to_vec();
+            let path = path.as_ref().to_path_buf();
+            rayon::spawn(move ||
+                // Errors? What Errors?
+                match fs::write(&path, &contents) {
+                    Ok(_) => (),
+                    Err(e) => {
+                        // In principle these should get displayed at the top
+                        // level, but just in case, send to stderr as well.
+                        eprintln!("\"{}\": {}", path.display(), e);
```

Comment:
Just like your comment stated, isn't it a duplicate?

Reply:
So I'm not 100% sure. I'd hate for a user to have a bad experience tracking down a problem because we failed to report an error reliably. The cost is perhaps showing the error twice. The benefit is increased surity it is delivered. A different code structure where e.g. we actually collected the results rather than tossing them into the void and letting the threadpool infra collect panics would be preferrable, but that will require considerably more work.

Ultimately though, I'm not going to get the bug reports from users that wonder why things are not reported, if in fact there are any reporting issues, so, if you want the eprintln! removed, just say so and I'll remove it.

Reply:
What about storing all those errors in a struct that is created outside the threads and just return from the current thread. Then print all the stored errors. Wouldn't it be a bit better?

Reply:
I don't really know enough about how and where the rustdoc internals are used to know what assumptions are safe to make in that regard. Is librustdoc used as a library, or is it really just the used by the CLI? Are there other entry points in, or is gathering in render appropriate? We could use an mpsc to gather feedback from threads and examine that in run()?

That could be done in a follow up patch too though, it doesn't need to be done in this iteration.

TBH, I'd like to hand this over to someone interested in maintaining the code: it was mainly a drive by test to see if the lessons learnt in rustup were applicable to rustdoc based on a twitter conversation: they appear to be so. I've polished it to the point where I believe that its correct - it should not violated any existing assumptions in the code (e.g. the flock requirements for shared file writing), and I believe that this patch as-is is unlikely to create new bug reports for the team). Equally I'm sure that  many more things can be done to it to make it much better - such as gathering the results of the IO calls into a mpsc queue or similar. Further profiling would also allow identifying how much of the remaining time is compiler performance and so on; a minute is a long time - unpacking the entire rust std docs is only 17 seconds or so on this machine with a very similar approach.

Reply:
I'll make a few changes on your branch then if you don't mind. :)

Reply:
Please do, I'd hate the effort of prototyping to be wasted, but you're in a much better position to assess the interactions raised here, unless I find a chunk of time I just don't have right now.